lib/spack/spack/spec.py

```diff
@@ -4069,6 +4105,58 @@ def do_parse(self):
 
         return specs
 
+    def _parse_dependency(self, specs):
+        """Parse the bit of a spec after a dependency sigil ^
+
+        Prerequisite is that the last accepted token was DEP.
+
+        Args:
+            specs: list of specs to be updated while parsing
+        """
+        if not specs:
+            # We're parsing an anonymous spec beginning with a
+            # dependency. Push the token to recover after creating
+            # anonymous spec
+            self.push_tokens([self.token])
+            specs.append(self.spec(None))
+        else:
+            dep = None
+            if self.accept(FILE):
+                # this may return None, in which case we backtrack
+                dep = self.spec_from_file()
+
+            if not dep and self.accept(HASH):
+                # We're finding a dependency by hash for an
+                # anonymous spec
+                dep = self.spec_by_hash()
+                dep = dep.copy(deps=('link', 'run'))
+
+            if not dep:
+                # We're adding a dependency to the last spec. This dependency
+                # can be either of the form ^spec i.e. a usual dependency or
+                # of the form ^vdep=spec if we have to bind a vdep to a
+                # particular provider.
+                self.expect(ID)
+                dep_or_virtual = self.token.value
+                if self.accept(EQ):
+                    # ^vdep=spec
+                    virtual_dependency = dep_or_virtual
+                    self.expect(VAL)
+                    # Implicit recursion on the parser below
+                    dep = Spec(self.token.value)
+                    specs[-1]._add_explicit_provider(virtual_dependency, dep)
```

Comment:
This is where we should be calling add_dependency and adding attributes to the edge.

Reply:
Unfortunately adding attributes here is not the only thing that's needed. The information parsed here is apparently lost and needs to be reconstructed every time we call:
```
self.flat_dependencies(copy=False)
```